- We started by opening and reading the files that contain the data, then, we listed them in a list of lists.


In [1]:
from csv import reader
from csv import reader

open_apple = open('AppleStore.csv', encoding='utf8')
open_google = open('googleplaystore.csv', encoding='utf8')

read_apple = reader(open_apple)
read_google = reader(open_google)

list_apple = list(read_apple)
list_google = list(read_google)

apple_header = list_apple[0]
apple = list_apple[1:]

google_header = list_google[0]
google = list_google[1:]
open_apple = open('AppleStore.csv', encoding='utf8')
open_google = open('googleplaystore.csv', encoding='utf8')

read_apple = reader(open_apple)
read_google = reader(open_google)

list_apple = list(read_apple)
list_google = list(read_google)

- In the next step, we created a function to make sure of the last step, and to explore what we just opened and listed.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

- Here we trying our function on the apple store apps.

In [3]:
explore_apple = explore_data(apple, 1, 4, True)

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1']


Number of rows: 7197
Number of columns: 16


- Trying the last step again, but, on google play apps.

In [4]:
explore_google = explore_data(google, 1, 4, True)

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 10841
Number of columns: 13


- Let's explore the column names to identify which columns could help us while we analyzing the apps. 

In [5]:
print(apple_header)
print('\n')
print(google_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


We see that the Google Play data set has 10841 apps and 13 columns. At a quick glance, the columns that might be useful for the purpose of our analysis are **App, Category, Reviews, Installs, Type, Price, and Genres.** Check the data set [documentation](https://www.kaggle.com/datasets/lava18/google-play-store-apps)

We have 7197 iOS apps in this data set, and the columns that seem interesting are: **track_name, currency, price, rating_count_tot, rating_count_ver, and prime_genre.** Not all column names are self-explanatory in this case, but details about each column can be found in the data set [documentation](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps)

# Cleaning Our Data


- we found an issue in the index #10472 in google apps data. The rating in that index was missed, so, we deleted that index by using del statement.

In [6]:
print(google[10472])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [7]:
del google[10472]

In [8]:
print(google[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


## Duplicated data
- There are a lot of duplicated data and the following code will show us an example of Instagram which is duplicated 4 times.

In [9]:
for app in google:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


## Duplicated list:
- As shown below, the duplicated list have 1181 duplicated application, which need to be removed.

In [10]:
duplicated = []
unique = []
for app in google:
    name = app[0]
    if name in unique:
        duplicated.append(name)
    elif name not in unique:
        unique.append(name)
print("All duplicated = ", len(duplicated),', examples:',duplicated[:15])

All duplicated =  1181 , examples: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


- We can use another criterion to remove the duplicated application, instead of using the application name, we can use the reviews count to keep the application that has the highest number of the reviews which will be the most updated one.

Also, we can use the last update column to keep the last update application and remove the duplicated applications which has the an old updated date.

# Remove the duplicate with another method
As we explained above, we can use another method to remove the duplicated application from our dataset, below, we will use the reviews count to remove the duplicated application.
If the reviews count was the highest, that means we have the last updated row of that application in the dataset, so, we will remove any other rows for the app with lower reviews.
- Below, we will:
- Create a dictionary where each key is a unique app name, and the value is the highest number of reviews of that app
- Use the dictionary to create a new data set, which will have only one entry per app (and we only select the apps with the highest number of reviews)

In [11]:
reviews_max = {}
for app in google:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
        
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


- After creating the dictionary, we can use it to create the new data set, so, below, we will:
- Start by initializing two empty lists, android_clean and already_added.
- Loop through the android data set, and for every iteration:
- We isolate the name of the app and the number of reviews.
- We add the current row (app) to the android_clean list, and the app name (name) to the already_added list if:
- The number of reviews of the current app matches the number of reviews of that app as described in the reviews_max dictionary; and The name of the app is not already in the already_added list. We need to add this supplementary condition to account for those cases where the highest number of reviews of a duplicate app is the same for more than one entry (for example, the Box app has three entries, and the number of reviews is the same). If we just check for reviews_max[name] == n_reviews, we'll still end up with duplicate entries for some apps.

In [12]:
android_clean = [] #our new cleaned data set
already_added = [] #it will just store the names of the apps

for app in google:
    name = app[0]
    n_reviews = float(app[3])
    
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)


print(len(android_clean))
print(android_clean[:3])

9659
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


# Checking the English apps only. Part 1.

In [13]:
def check_names(our_string):
    for chara in our_string:
        name = ord(chara)
        if name >= 127:
            return False
    return True

check_names('Instagram')
check_names('爱奇艺PPS -《欢乐颂2》电视剧热播')
check_names('Docs To Go™ Free Office Suite')
check_names('Instachat 😜')      

False

# Checking the English apps only. Part 2.

In [14]:
def check_namess(our_string):
    count_chara = 0
    
    for chara in our_string:
        name = ord(chara)
        if name >= 127:
            count_chara += 1
            
    if count_chara > 3:
        return False
    
    return True

check_namess('Instagram')
check_namess('爱奇艺PPS -《欢乐颂2》电视剧热播')
check_namess('Docs To Go™ Free Office Suite')
check_namess('Instachat 😜')  

True

# Apply the previous function on our data set

In [15]:
english_google = []
english_apple = []

for app in android_clean:
    if check_namess(app[0]):
        english_google.append(app)
        
print(english_google[:3])
print('\n')
print(len(english_google)) 


for app in apple:
    if check_namess(app[1]):
        english_apple.append(app)
        
print('\n')
print(english_apple[:3])
print('\n')
print(len(english_apple)) 

[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']]


9614


[['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']]


6183


# Isolating the Free Apps

In [16]:
free_google = []
non_free_google = []

free_apple = []
non_free_apple = []

for app in english_google:
    free = app[7]
    
    if free != '0':
        non_free_google.append(app)
    else:
        free_google.append(app)
        
for app in english_apple:
    price = app[4]
    
    if price != '0.0':
        non_free_apple.append(app)
    else:
        free_apple.append(app)
        
print(len(free_google))
print(len(free_apple))


8864
3222


# Most Common Apps by Genre: Part One
- In the following steps, we will find the most common apps genre in both google play and apple store.
- Finding the most common apps genre will help us to determine the nature of the app that we need to create to reach to the most users to increase our revenue. 
- Let's beging by creating frequency tables for some columns to find the most applications have the same genres.

In [17]:
print(google_header)
print('\n')
print(apple_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


- It seems that we need to make frequency tables for the category, rating, reviews and genres for google play data set.
- For Apple Store, we need to make frequency tables for 'rating_count_tot', 'user_rating' and 'prime_genre'.

In [18]:
google_genres = []
google_category = []
apple_genre = []

for app in free_google:
    genre = app[9]
    if genre not in google_genres:
        google_genres.append(genre)
for app in free_google:
    category = app[1]
    if category not in google_category:
        google_category.append(category)

for app in free_apple:
    genre = app[11]
    if genre not in apple_genre:
        apple_genre.append(genre)
        
#print(google_genres)
#print('\n')
#print(google_category)
#print('\n')
#print(apple_genre)

In [19]:
google_genre_dic = {}
google_categ_dic = {}
apple_genre_dic  = {}

for genre in google_genres:
    google_genre_dic[genre] = 0
for categ in google_category:
    google_categ_dic[categ] = 0
for genre in apple_genre:
    apple_genre_dic[genre] = 0
    
#print(google_genre_dic)
#print('\n')
#print(google_categ_dic)
#print('\n')
#print(apple_genre_dic)

In [20]:
for app in free_google:
    genre = app[9]
    
    for name in google_genre_dic:
        if name == genre:
            google_genre_dic[name] += 1
            
            
for app in free_google:
    category = app[1]
    
    for name in google_categ_dic:
        if name == category:
            google_categ_dic[name] += 1
            

for app in free_apple:
    genre = app[11]
    
    for name in apple_genre_dic:
        if name == genre:
            apple_genre_dic[name] += 1
            
#print(google_genre_dic)
#print('\n')
#print(google_categ_dic)
#print('\n')
print(apple_genre_dic)

{'Social Networking': 106, 'Photo & Video': 160, 'Games': 1874, 'Music': 66, 'Reference': 18, 'Health & Fitness': 65, 'Weather': 28, 'Utilities': 81, 'Travel': 40, 'Shopping': 84, 'News': 43, 'Navigation': 6, 'Lifestyle': 51, 'Entertainment': 254, 'Food & Drink': 26, 'Sports': 69, 'Book': 14, 'Finance': 36, 'Education': 118, 'Productivity': 56, 'Business': 17, 'Catalogs': 4, 'Medical': 6}


# Let's try another method by creating a function

In [21]:
def genras(data_set, index):
    our_frequency_table = {}
    
    for app in data_set:
        name = app[index]
        
        if name in our_frequency_table:
            our_frequency_table[name] += 1
        elif name not in our_frequency_table:
            our_frequency_table[name] = 1 
            
    return our_frequency_table

print(genras(free_apple, 11))

{'Social Networking': 106, 'Photo & Video': 160, 'Games': 1874, 'Music': 66, 'Reference': 18, 'Health & Fitness': 65, 'Weather': 28, 'Utilities': 81, 'Travel': 40, 'Shopping': 84, 'News': 43, 'Navigation': 6, 'Lifestyle': 51, 'Entertainment': 254, 'Food & Drink': 26, 'Sports': 69, 'Book': 14, 'Finance': 36, 'Education': 118, 'Productivity': 56, 'Business': 17, 'Catalogs': 4, 'Medical': 6}


# Let's display our function with percentage instead of numbers.

In [22]:
def genras(data_set, index):
    our_frequency_table = {}
    total = 0
    for app in data_set:
        name = app[index]
        total += 1
        if name in our_frequency_table:
            our_frequency_table[name] += 1
        elif name not in our_frequency_table:
            our_frequency_table[name] = 1
            
    percentage_table = {}
    
    for entry in our_frequency_table:
        percentage = (our_frequency_table[entry] / total) * 100
        percentage_table[entry] = percentage
        
    return percentage_table

genras(free_apple, 11)


{'Social Networking': 3.2898820608317814,
 'Photo & Video': 4.9658597144630665,
 'Games': 58.16263190564867,
 'Music': 2.0484171322160147,
 'Reference': 0.5586592178770949,
 'Health & Fitness': 2.0173805090006205,
 'Weather': 0.8690254500310366,
 'Utilities': 2.5139664804469275,
 'Travel': 1.2414649286157666,
 'Shopping': 2.60707635009311,
 'News': 1.3345747982619491,
 'Navigation': 0.186219739292365,
 'Lifestyle': 1.5828677839851024,
 'Entertainment': 7.883302296710118,
 'Food & Drink': 0.8069522036002483,
 'Sports': 2.1415270018621975,
 'Book': 0.4345127250155183,
 'Finance': 1.1173184357541899,
 'Education': 3.662321539416512,
 'Productivity': 1.7380509000620732,
 'Business': 0.5276225946617008,
 'Catalogs': 0.12414649286157665,
 'Medical': 0.186219739292365}

# We can use another function to sort the result of the last function

In [23]:
def display_table(dataset, index):
    freq_table = genras(dataset, index)
    our_new_list_of_tuple = []
    for key in freq_table:
        our_tuple = (freq_table[key], key)
        our_new_list_of_tuple.append(our_tuple)
        
    sorted_data = sorted(our_new_list_of_tuple, reverse = True)
    
    for entry in sorted_data:
        print(entry[1], ":", entry[0])
        
#display_table(free_apple, 11)
#print('\n')
#display_table(free_google, 1)
#print('\n')
#display_table(free_google, 9)

# Most Popular Apps by Genre on the App Store

In [24]:
our_apple_genre = genras(free_apple, 11)

for genre in our_apple_genre:
    total = 0
    len_genre = 0
    
    for app in free_apple:
        genre_app = app[11]
        
        if genre_app == genre:
            
            rating = float(app[5])
            total += rating
            len_genre += 1
    
    average_rating = total / len_genre

    print(genre, ':', average_rating)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


# Most Popular Apps by Genre on Google Play

In [25]:
our_google_category = genras(free_google, 1)

for app in our_google_category:
    total = 0
    len_cat = 0
    
    for category in free_google:
        catt = category[1]
        
        if catt == app:
            installs = category[5]
            installs = installs.replace('+','')
            installs = installs.replace(',','')
            installs = float(installs)
            total += installs
            len_cat += 1
            
    average_caty = total/len_cat
    print(app, ':', average_caty)


ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_